terraform/main.tf

```diff
@@ -204,6 +248,242 @@ resource "azurerm_federated_identity_credential" "service_operator" {
+    allowed_member_types = ["User"]
+    description          = "Allows the app to read the memberships of all groups."
+    display_name         = "GroupMember.Read.All"
+    value                = "GroupMember.Read.All"
+  }
+
+  required_resource_access {
+    resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph API
+
+    resource_access {
+      id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read
+      type = "Scope"
+    }
+
+    resource_access {
+      id   = "df021288-bdef-4463-88db-98f22de89214" # User.Read.All
+      type = "Role"
+    }
+
+    resource_access {
+      id   = "98830695-27a2-44f7-8c18-0c3ebc9698f6" # GroupMember.Read.All
+      type = "Role"
+    }
+
+    resource_access {
+      id   = "64a6cdd6-aab1-4aaf-94b8-3cc8405e90d0" # email
+      type = "Scope"
+    }
+
+    resource_access {
+      id   = "7427e0e9-2fba-42fe-b0c0-848c9e6a8182" # offline_access
+      type = "Scope"
+    }
+
+    resource_access {
+      id = "e383f46e-2787-4529-855e-0e479a3ffac0" # mail.send
+      type = "Scope"
+    }
+
+    resource_access {
+      id   = "37f7f235-527c-4136-accd-4a02d197296e" # openid
+      type = "Scope"
+    }
+  }
+}
+
+
+# Define the OAuth2 permissions (redirect URIs)
+resource "azuread_application_redirect_uris" "backstage_redirect_uri" {
+  count = local.build_backstage ? 1 : 0
+  application_id = "/applications/${azuread_application.backstage-app[count.index].object_id}"
+  type                  = "Web"
+  redirect_uris         = ["https://${azurerm_public_ip.backstage_public_ip[count.index].ip_address}/api/auth/microsoft/handler/frame"]
+}
+# Define the service principal
+resource "azuread_service_principal" "backstage-app-sp" {
+  count = local.build_backstage ? 1 : 0
+  client_id = azuread_application.backstage-app[count.index].application_id
+}
+
+# Define the service principal password
+resource "azuread_service_principal_password" "backstage-sp-password" {
+  count = local.build_backstage ? 1 : 0
+  service_principal_id = azuread_service_principal.backstage-app-sp[count.index].id
+  end_date             = "2099-01-01T00:00:00Z"
+}
+
+resource "null_resource" "ascii_art" {
+  count = local.build_backstage ? 1 : 0
+  depends_on = [ azuread_service_principal_password.backstage-sp-password ]
+  provisioner "local-exec" {
+    command = <<EOT
+echo "    _      _     ______  _____   _______ "
+echo "   / \\   | |   |  ____||  __ \\|__   __|"
+echo "  / _ \\  | |   | |__   | |__) |   | |   "
+echo " / ___ \\ | |   |  __|  |  _  /    | |   "
+echo "/_/   \\_\|_|___| |____ | | \\\    | |   "
+echo "        \\_\\_____|______||_| \_\  |_|   "
+echo ""
+echo ""
+echo "Please grant admin consent on app registration now to avoid waiting for the 1 hour schedule post backstage chart deployment."
+EOT
+  }
+}
+
+# Output the necessary variables
+output "azure_client_id" {
+value = length(azuread_application.backstage-app) > 0 ? azuread_application.backstage-app[0].application_id : null
```

Comment:
Not going to block PR on this, but would be nice to get rid of these new warning messages:

```
│ Warning: Deprecated attribute
│ 
│   on main.tf line 363, in output "azure_client_id":
│  363: value = length(azuread_application.backstage-app) > 0 ? azuread_application.backstage-app[0].application_id : null
│ 
│ The attribute "application_id" is deprecated. Refer to the provider documentation for details.
│ 
│ (and 2 more similar warnings elsewhere)
```

Reply:
Will pick up when I have some more time. Agreed warnings are a little ugly and long term will need to change although some way away in case of app id.